plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -79,97 +82,124 @@ Then, in the _body_, we track the term and type application (`Apply` or `TyInst`
+-- being applied and the arguments to which it is applied
+collectArgs :: Term tyname name uni fun ann
+  -> (Term tyname name uni fun ann, ArgOrderWithAnn tyname name uni fun ann)
+collectArgs expr
+  = go expr []
+  where
+    go (Apply ann f a) as      = go f ((MkTermArg a, ann):as)
+    go (TyInst ann f tyArg) as = go f ((MkTypeArg tyArg, ann):as)
+    go e as                    = (e, as)
+
+-- | Apply a list of term and type arguments to a function in potentially a nested fashion.
+mkApps :: Term tyname name uni fun ann
+  -> ArgOrderWithAnn tyname name uni fun ann
+  -> Term tyname name uni fun ann
+mkApps f ((MkTermArg tmArg, ann) : args) = mkApps (Apply ann f tmArg) args
+mkApps f ((MkTypeArg tyArg, ann) : args) = mkApps (TyInst ann f tyArg) args
+mkApps f []                              = f
+
+-- | Given the arity of a function, and the list of arguments applied to it, return whether it is
+-- fully applied or not.
+isFullyApplied :: Arity -> ArgOrder tyname name uni fun ann -> Bool
+isFullyApplied [] (_argsOrder:_as) = True -- over-application
+isFullyApplied (_arity:_) [] = False -- under-application
+isFullyApplied [] [] = True
+isFullyApplied lamOrder argsOrder =
+  -- start comparing from the end because there may be over-application
+  case (last lamOrder, last argsOrder) of
+    (MkTerm, MkTermArg _) -> isFullyApplied (init lamOrder) (init argsOrder)
+    (MkType, MkTypeArg _) -> isFullyApplied (init lamOrder) (init argsOrder)
+    _                     -> False
+
+-- | Inline fully applied functions iff the body of the function is `acceptable`.
+inlineSat :: forall tyname name uni fun ann. InliningConstraints tyname name uni fun
+    => Term tyname name uni fun ann -- ^ The `body` of the `Let` term.
+    -> InlineM tyname name uni fun ann (Term tyname name uni fun ann)
+inlineSat appOrTyInstTm =
+    case appOrTyInstTm of
+      -- If the term is a term or type application, check it's applying to a var that we may inline
+      Apply _varAnn _fun _arg  -> go appOrTyInstTm
+      TyInst _varAnn _fun _arg -> go appOrTyInstTm
+      -- otherwise, check all subterms
+      _                        -> forMOf termSubterms appOrTyInstTm inlineSat
+      where
+        go tm = do
+            -- collect all the arguments of the term being applied to
+            let argsAppliedTo = fst $ collectArgs tm
+                args = snd $ collectArgs tm
+            case argsAppliedTo of
+              -- if it is a `Var` that is being applied to, check to see if it's fully applied
+              Var _ann name -> do
+                maybeVarInfo <- gets (lookupCalled name)
+                case maybeVarInfo of
+                  -- the variable maybe a *recursive* let binding, in which case we don't process,
+                  -- and it won't be in the map. ATM recursive bindings aren't inlined.
+                  Nothing -> pure tm
+                  Just varInfo -> do
+                    isAcceptable <- acceptable (calledVarBody varInfo)
```

Comment:
Hmm, why is `acceptable` monadic? It shouldn't be, looking at the definition.

Reply:
good point. Fixed.